File: hiob_contracts/planets/ares/script_generation_v1.py
"""Exact provider input emitted by ``ares.script.prepare_generation``."""

from __future__ import annotations

from typing import Annotated, Any, Literal, Mapping

from pydantic import (
    AfterValidator,
    BaseModel,
    ConfigDict,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)

from ...ares_script_revision_v1 import (
    _FROZEN_STRICT,
)
from ...character_identity_v1 import (
    character_identity_binding_errors_v1,
)
from ...brand_scope import is_contract_blank
from ...factory import sha256_digest
from ...voice_spec_v1 import derive_voice_spec_digest_v1


_DIGEST_FIELDS = (
    "contract_version",
    "workspace_id",
    "run_id",
    "script_revision_id",
    "plan_revision_id",
    "factory_revision",
    "character_lock",
    "voice_spec",
    "current_character",
    "conflict",
    "adjacent_beat_summaries",
    "memories",
)


def _valid_unicode_scalars(value: str) -> str:
    if any(0xD800 <= ord(character) <= 0xDFFF for character in value):
        raise ValueError("text must contain valid Unicode scalar values")
    return value


def _contract_nonblank(value: str) -> str:
    if is_contract_blank(value):
        raise ValueError("string must not be blank")
    return value


def _assert_json_unicode_scalars(value: Any) -> None:
    if isinstance(value, str):
        _valid_unicode_scalars(value)
    elif isinstance(value, Mapping):
        for key, item in value.items():
            _valid_unicode_scalars(str(key))
            _assert_json_unicode_scalars(item)
    elif isinstance(value, (list, tuple)):
        for item in value:
            _assert_json_unicode_scalars(item)


_CONTRACT_NONBLANK_PATTERN = (
    "[^\u0009-\u000D\u0020\u00A0\u1680"
    "\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF]"
)

Text80 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=80,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
Text120 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=120,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
Text128 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=128,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
Text200 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=200,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
Text256 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=256,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
Text300 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=300,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
Text500 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=500,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
Text512 = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=512,
        pattern=_CONTRACT_NONBLANK_PATTERN,
    ),
    AfterValidator(_contract_nonblank),
    AfterValidator(_valid_unicode_scalars),
]
DigestText = Annotated[
    str,
    StringConstraints(pattern=r"^sha256:[0-9a-f]{64}$"),
]
_FACTORY_REVISION_INTEGER_ERROR = "factory_revision must be an integer"


def _normalize_factory_revision(value: Any) -> Any:
    if isinstance(value, bool):
        raise ValueError(_FACTORY_REVISION_INTEGER_ERROR)
    if isinstance(value, float):
        if not value.is_integer():
            raise ValueError(_FACTORY_REVISION_INTEGER_ERROR)
        value = int(value)
    if not isinstance(value, int):
        raise ValueError(_FACTORY_REVISION_INTEGER_ERROR)
    if not 0 <= value <= 2_147_483_647:
        raise ValueError("factory_revision must be an int4 value")
    return value


def derive_ares_script_generation_input_digest_v1(
    value: Mapping[str, Any] | BaseModel,
) -> str:
    """Bind the complete bounded context and exclude only its own digest."""

    data = (
        value.model_dump(mode="json")
        if isinstance(value, BaseModel)
        else dict(value)
    )
    for field in _DIGEST_FIELDS:
        if field not in data:
            raise ValueError(
                f"{field} is required for generation input digest"
            )
    data["factory_revision"] = _normalize_factory_revision(
        data["factory_revision"]
    )
    body = {field: data[field] for field in _DIGEST_FIELDS}
    _assert_json_unicode_scalars(body)
    return sha256_digest(body)


def _semantic_model_config(*invariants: str) -> ConfigDict:
    return ConfigDict(
        **_FROZEN_STRICT,
        json_schema_extra={
            "x-hiob-validation": "pydantic-runtime-required",
            "x-hiob-semantic-invariants": list(invariants),
        },
    )


class AresCharacterIdentityProjectionV1(BaseModel):
    """The face and voice identity projection Ares is allowed to consume."""

    model_config = _semantic_model_config(
        "character_identity_binding_digest",
        "valid_unicode_scalars",
    )

    persona_id: Text128
    face_id: Text256
    voice_id: Text256
    identity_binding_digest: DigestText

    @model_validator(mode="after")
    def _bind_face_and_voice(self) -> "AresCharacterIdentityProjectionV1":
        errors = character_identity_binding_errors_v1(
            subject_id=self.persona_id,
            face_id=self.face_id,
            voice_id=self.voice_id,
            identity_binding_digest=self.identity_binding_digest,
        )
        if errors:
            raise ValueError(errors[0])
        return self


class AresProvenanceMemoryV1(BaseModel):
    """One bounded memory whose source remains visible."""

    model_config = _semantic_model_config("valid_unicode_scalars")

    text: Text500
    provenance: Text200


class AresVoiceSpecProjectionV1(BaseModel):
    """The exact VoiceSpec fields Ares passes to its provider."""

    model_config = _semantic_model_config(
        "voice_spec_digest",
        "valid_unicode_scalars",
    )

    contract_version: Literal["VoiceSpec.v1"]
    subject_id: Text128
    rhythm: Text300
    vocabulary: tuple[Text80, ...] = Field(max_length=12)
    forbidden_phrases: tuple[Text120, ...] = Field(max_length=12)
    approved_examples: tuple[Text500, ...] = Field(
        min_length=3,
        max_length=5,
    )
    voice_spec_digest: DigestText

    @field_validator(
        "vocabulary",
        "forbidden_phrases",
        "approved_examples",
        mode="before",
    )
    @classmethod
    def _tuples(cls, value: Any) -> Any:
        return tuple(value) if isinstance(value, list) else value

    @model_validator(mode="after")
    def _bind_voice_spec(self) -> "AresVoiceSpecProjectionV1":
        if self.voice_spec_digest != derive_voice_spec_digest_v1(self):
            raise ValueError(
                "voice_spec_digest does not match VoiceSpec content"
            )
        return self


class AresScriptGenerationInputV1(BaseModel):
    """The only payload Ares permits the script provider to receive."""

    model_config = _semantic_model_config(
        "character_identity_binding_digest",
        "voice_spec_subject_matches_character",
        "voice_spec_digest",
        "generation_input_digest",
        "valid_unicode_scalars",
    )

    contract_version: Literal["AresScriptGenerationInput.v1"]
    workspace_id: Text512
    run_id: Text512
    script_revision_id: Text512
    plan_revision_id: Text512
    factory_revision: int = Field(
        ge=0,
        le=2_147_483_647,
        strict=True,
    )
    character_lock: AresCharacterIdentityProjectionV1
    voice_spec: AresVoiceSpecProjectionV1
    current_character: Text500
    conflict: Text500
    adjacent_beat_summaries: tuple[Text300, ...] = Field(max_length=2)
    memories: tuple[AresProvenanceMemoryV1, ...] = Field(
        max_length=3,
    )
    generation_input_digest: DigestText

    @field_validator("factory_revision", mode="before")
    @classmethod
    def _json_integer_parity(cls, value: Any) -> Any:
        return _normalize_factory_revision(value)

    @field_validator(
        "adjacent_beat_summaries",
        "memories",
        mode="before",
    )
    @classmethod
    def _tuples(cls, value: Any) -> Any:
        return tuple(value) if isinstance(value, list) else value

    def model_copy(
        self,
        *,
        update: Mapping[str, Any] | None = None,
        deep: bool = False,
    ) -> "AresScriptGenerationInputV1":
        if not update:
            return super().model_copy(deep=deep)
        body = self.model_dump(mode="json")
        body.update(update)
        return type(self).model_validate(body, strict=True)

    @model_validator(mode="after")
    def _bind_scope_and_digest(self) -> "AresScriptGenerationInputV1":
        if self.voice_spec.subject_id != self.character_lock.persona_id:
            raise ValueError(
                "voice_spec.subject_id must match character_lock.persona_id"
            )
        if (
            self.generation_input_digest
            != derive_ares_script_generation_input_digest_v1(self)
        ):
            raise ValueError(
                "generation_input_digest does not match generation input"
            )
        return self


__all__ = [
    "AresCharacterIdentityProjectionV1",
    "AresProvenanceMemoryV1",
    "AresVoiceSpecProjectionV1",
    "AresScriptGenerationInputV1",
    "derive_ares_script_generation_input_digest_v1",
]

File: ts/src/planets/ares/script-generation-v1.ts
/** Exact provider input emitted by `ares.script.prepare_generation`. */
import { z } from 'zod';

import {
  characterIdentityBindingErrorV1,
} from '../../character-identity-v1.js';
import { sha256Digest } from '../../factory/digest.js';
import { deriveVoiceSpecDigestV1 } from '../../voice-spec-v1.js';

const DigestSchema = z
  .string()
  .regex(/^sha256:[0-9a-f]{64}$/, 'digest must be sha256:<64 lowercase hex>');
function hasUnpairedSurrogate(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      const low = value.charCodeAt(index + 1);
      if (!(low >= 0xdc00 && low <= 0xdfff)) return true;
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return true;
    }
  }
  return false;
}

function assertJsonUnicodeScalars(value: unknown): void {
  if (typeof value === 'string') {
    if (hasUnpairedSurrogate(value)) {
      throw new TypeError('text must contain valid Unicode scalar values');
    }
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) assertJsonUnicodeScalars(item);
    return;
  }
  if (value !== null && typeof value === 'object') {
    for (const [key, item] of Object.entries(
      value as Record<string, unknown>,
    )) {
      if (hasUnpairedSurrogate(key)) {
        throw new TypeError('text must contain valid Unicode scalar values');
      }
      assertJsonUnicodeScalars(item);
    }
  }
}

const boundedNonBlankString = (maxLength: number) =>
  z
    .string()
    .refine(
      (value) => value.trim().length > 0,
      'string must not be blank',
    )
    .refine(
      (value) => !hasUnpairedSurrogate(value),
      'text must contain valid Unicode scalar values',
    )
    .refine(
      (value) => [...value].length <= maxLength,
      `string must contain at most ${maxLength} Unicode scalars`,
    );
const NonBlankString = boundedNonBlankString(512);

const DIGEST_FIELDS = [
  'contract_version',
  'workspace_id',
  'run_id',
  'script_revision_id',
  'plan_revision_id',
  'factory_revision',
  'character_lock',
  'voice_spec',
  'current_character',
  'conflict',
  'adjacent_beat_summaries',
  'memories',
] as const;

function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === 'object') {
    for (const child of Object.values(value as Record<string, unknown>)) {
      deepFreeze(child);
    }
    Object.freeze(value);
  }
  return value;
}

export function deriveAresScriptGenerationInputDigestV1(
  value: Record<string, unknown>,
): string {
  const body = Object.fromEntries(
    DIGEST_FIELDS.map((field) => {
      if (!(field in value) || value[field] === undefined) {
        throw new TypeError(`${field} is required for generation input digest`);
      }
      return [field, value[field]];
    }),
  );
  assertJsonUnicodeScalars(body);
  return sha256Digest(body);
}

export const AresCharacterIdentityProjectionV1Schema = z
  .object({
    persona_id: boundedNonBlankString(128),
    face_id: boundedNonBlankString(256),
    voice_id: boundedNonBlankString(256),
    identity_binding_digest: DigestSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    const error = characterIdentityBindingErrorV1({
      subject_id: value.persona_id,
      face_id: value.face_id,
      voice_id: value.voice_id,
      identity_binding_digest: value.identity_binding_digest,
    });
    if (error) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: error,
        path: ['identity_binding_digest'],
      });
    }
  });

export const AresProvenanceMemoryV1Schema = z
  .object({
    text: boundedNonBlankString(500),
    provenance: boundedNonBlankString(200),
  })
  .strict();

export const AresVoiceSpecProjectionV1Schema = z
  .object({
    contract_version: z.literal('VoiceSpec.v1'),
    subject_id: boundedNonBlankString(128),
    rhythm: boundedNonBlankString(300),
    vocabulary: z.array(boundedNonBlankString(80)).max(12),
    forbidden_phrases: z.array(boundedNonBlankString(120)).max(12),
    approved_examples: z
      .array(boundedNonBlankString(500))
      .min(3)
      .max(5),
    voice_spec_digest: DigestSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.voice_spec_digest !== deriveVoiceSpecDigestV1(value)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'voice_spec_digest does not match VoiceSpec content',
        path: ['voice_spec_digest'],
      });
    }
  });

export const AresScriptGenerationInputV1Schema = z
  .object({
    contract_version: z.literal('AresScriptGenerationInput.v1'),
    workspace_id: NonBlankString,
    run_id: NonBlankString,
    script_revision_id: NonBlankString,
    plan_revision_id: NonBlankString,
    factory_revision: z.number().int().min(0).max(2_147_483_647),
    character_lock: AresCharacterIdentityProjectionV1Schema,
    voice_spec: AresVoiceSpecProjectionV1Schema,
    current_character: boundedNonBlankString(500),
    conflict: boundedNonBlankString(500),
    adjacent_beat_summaries: z
      .array(boundedNonBlankString(300))
      .max(2),
    memories: z
      .array(AresProvenanceMemoryV1Schema)
      .max(3),
    generation_input_digest: DigestSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.voice_spec.subject_id !== value.character_lock.persona_id) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message:
          'voice_spec.subject_id must match character_lock.persona_id',
        path: ['voice_spec', 'subject_id'],
      });
    }
    let expectedDigest: string | null = null;
    try {
      expectedDigest = deriveAresScriptGenerationInputDigestV1(value);
    } catch {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'generation input must contain valid Unicode scalar values',
        path: ['generation_input_digest'],
      });
      return;
    }
    if (value.generation_input_digest !== expectedDigest) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'generation_input_digest does not match generation input',
        path: ['generation_input_digest'],
      });
    }
  })
  .transform(deepFreeze);

export type AresCharacterIdentityProjectionV1 = z.infer<
  typeof AresCharacterIdentityProjectionV1Schema
>;
export type AresProvenanceMemoryV1 = z.infer<
  typeof AresProvenanceMemoryV1Schema
>;
export type AresVoiceSpecProjectionV1 = z.infer<
  typeof AresVoiceSpecProjectionV1Schema
>;
export type AresScriptGenerationInputV1 = z.infer<
  typeof AresScriptGenerationInputV1Schema
>;
